### PR TITLE
July 2021 feature set

### PR DESCRIPTION
**Pull request by alexpreynolds**
- UCSC genome browser-like annotation block style
- UCSC genome browser-like intron, exon, 5'UTR, 3'UTR feature count labeling
- Boxplot (Tukey box-and-whisker) annotation block style
- Color applied from BED12 annotation field (unless overwritten by user-specified options, or unspecified and determined by strand, if available)
- Canonical or APPRIS isoform highlight flag and color options
- Improved PNG/SVG export for all `blockStyle` modes
- Can now specify maximum row rendering for constrained device viewport (i.e., mobile)

Some example animations of block style modifications:

1. Feature counts and custom coloring:

Screen Recording 2021-07-14 at 4 13 41 PM

2. UCSC genome browser-like block style:

Screen Recording 2021-07-14 at 4 20 44 PM

3. Tukey boxplot block style:

Screen Recording 2021-07-14 at 4 24 20 PM

4. Canonical isoform highlighting:

Screen Shot 2021-07-14 at 4 45 11 PM

### CODE REVIEW

**Comment by alexpreynolds** [comment]
Just wanted to check in on this.

**Comment by alexander-veit** [comment]
I will have a look today. Sorry it took so long.

File: README.md [hunk 1]
```diff
@@ -83,9 +83,16 @@ The following options are available:
...
+    "maxRows": null, // Maximum number of rows to display (useful for mobile devices)
     "showToggleTranscriptsButton": true, // If the "Show fewer transcripts"/"Show more transcripts" is shown
     "trackHeightAdjustment": "automatic", // if "automatic", the height of the track is adjusted to the number of visible transcripts.
     "startCollapsed": false, // if true, only one transcript is shown
+    "blockStyle": "directional", // "UCSC-like" for UCSC block style, or "boxplot" to use BED12 attributes to mimic Tukey box-and-whisker plot style 
+    "highlightTranscriptType": "none", // If "longestIsoform" or "apprisPrincipalIsoform", a principal or canonical annotation is highlighted.
+    "highlightTranscriptTrackBackgroundColor": "#f0f0f0",
+    "highlightTranscriptLabelBackgroundColor": "#f0f0f0",
+    "highlightTranscriptLabelFontWeight": "700",
+    "showHighlightedTranscriptsOnly": false, // Only show highlighted annotations
```
Comment (alexander-veit) [inline README.md:100]: Add `isVisible`, `trackMargin` and `allowResizeParentDiv` to the README

File: src/scripts/TranscriptsTrack.js [hunk 2]
```diff
@@ -1719,38 +3348,60 @@ TranscriptsTrack.config = {
...
     "trackHeightAdjustment",
     "sequenceData",
     "backgroundColor",
+    "blockStyle",
+    "highlightTranscriptType",
+    "highlightTranscriptTrackBackgroundColor",
+    "highlightTranscriptLabelBackgroundColor",
+    "highlightTranscriptLabelFontWeight",
+    "showHighlightedTranscriptsOnly",
+    "isVisible",
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js:3420]: When `isVisible` is `true`, the "Show fewer transcripts" button is still shown. In this case the overall track height is not 0.
Also, we might want to add
`if (!this.options.isVisible) return "";` to `getMouseoverHtml`.

File: src/scripts/TranscriptsTrack.js [hunk 3]
```diff
@@ -389,58 +532,148 @@ const TranscriptsTrack = (HGC, ...args) => {
...
+            graphics.drawPolygon(chevronPoly);
+            graphics.endFill();
+            polysSVG.push({
+              rect: chevronPoly,
+              color: (itemRgbIndex !== -1) ? colorUsedSVG : track.colors.intronHEX,
+              colorMode: itemRgbColorMode,
+              paintOrder: 0
+            });
+          }
+        }
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js]: See below

File: src/scripts/TranscriptsTrack.js [hunk 4]
```diff
@@ -389,58 +532,148 @@ const TranscriptsTrack = (HGC, ...args) => {
...
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js]: I would prefer to not add the chevrons for the standard 'directional' case as it adds some complexity to an already expensive track that might not be needed in this case. Can you execute this block only when the blockStyle is not directional?

File: src/scripts/TranscriptsTrack.js [hunk 5]
```diff
@@ -670,19 +1131,25 @@ const TranscriptsTrack = (HGC, ...args) => {
...
+      this.colors["highlightLabelBackground"] = colorToHex(this.options.highlightTranscriptLabelBackgroundColor);
+      this.colors["highlightTrackBackground"] = colorToHex(this.options.highlightTranscriptTrackBackgroundColor);
     }
 
     initTile(tile) {
       externalInitTile(this, tile, {
         flipText: this.flipText,
-        fontSize: this.fontSize,
         fontFamily: this.options.fontFamily,
+        labelFontSize: this.options.labelFontSize,
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js:1146]: Could you remove the old `fontSize` option and replace it with `labelFontSize` everywhere please?

File: src/scripts/TranscriptsTrack.js [hunk 6]
```diff
@@ -726,45 +1199,757 @@ const TranscriptsTrack = (HGC, ...args) => {
...
+      let boxBlockStarts = [];
+      if (this.options.blockStyle === "boxplot") {
+        if (!startCodonPos) { startCodonPos = +ts[6] }
+        if (!stopCodonPos) { stopCodonPos = +ts[7] }
+        boxBlockCount = +ts[9];
+        boxBlockLengths = ts[10].split(",").map((x) => +x);
+        boxBlockStarts = ts[11].split(",").map((x) => +x);
+      }
+
+      function calculateBlocks() {
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js]: Oh boy... I didn't check this function line by line and assume you got it right :-)

File: src/scripts/TranscriptsTrack.js [hunk 7]
```diff
@@ -1349,7 +2661,49 @@ const TranscriptsTrack = (HGC, ...args) => {
...
+          //this.renderDirectionlessTextBg(this.allBoxes, this.allTexts, allTiles);
+          break;
+        default:
+          throw new Error(
+            'Uncaught TypeError: Unknown blockStyle option (drawLabels, within tile data, B)'
+          );
+      }
+    }
+
+    renderDirectionlessTextBg(allBoxes, allTexts, allTiles) {
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js:2706]: Looks like this isn't used?

File: src/scripts/TranscriptsTrack.js [hunk 8]
```diff
@@ -1428,8 +2782,8 @@ const TranscriptsTrack = (HGC, ...args) => {
 
     setDimensions(newDimensions) {
       this.updateTranscriptInfo();
+      // This will rerender all tiles.
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js:2812]: 👍

File: src/scripts/TranscriptsTrack.js [hunk 9]
```diff
@@ -1459,69 +2959,156 @@ const TranscriptsTrack = (HGC, ...args) => {
...
+                    itemRGB: (transcript.itemRgb !== '.') ? transcript.itemRgb : '0,0,0',
+                    blockCount: transcript.boxBlockCount,
+                    blockSizes: transcript.boxBlockLengths,
+                    blockStarts: transcript.boxBlockStarts,
+                  };
+
+                  const bed12Output = this.formattedBED12HTML(bed12FieldsObj);
+
+                  if (bed12Output.length === 0) {
+                    const hc = document.getElementsByClassName('higlass-main-content')[0];
```
Comment (alexander-veit) [inline src/scripts/TranscriptsTrack.js:3142]: I guess that only works, when there is only one Higlass display on the page. It's fine with me. I was just wondering if there is a simple way to get the "current" display.

**Review by alexander-veit — APPROVED** [review]
This is fantastic. Thanks a lot for these additions and improvements. That was a lot of work! 
It looks very good, just a few remarks below.

**Comment by alexpreynolds** [comment]
Thanks — will work through these and follow up.

File: README.md (hunk 1 above)
Comment (alexpreynolds) [reply README.md:100]: Added `isVisible` and `trackMargin` to `README` and `TranscriptsTrack.js` defaults in commit b6c5325

Comment (alexpreynolds) [reply README.md:100]: I removed `allowResizeParentDiv` as I do not provide functionality for it in the client apps that use this track type.

File: src/scripts/TranscriptsTrack.js (hunk 2 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js:3420]: Resolved in commit a87b8e8

File: src/scripts/TranscriptsTrack.js (hunk 5 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js:1146]: Resolved in commit b1aa091

Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js:1146]: I separated out the `label*` and `codon*` properties for their respective text style attributes.

File: src/scripts/TranscriptsTrack.js (hunk 6 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js]: I used Gencode hg19, hg38, and mm10 annotations to examine all the different scenarios I could picture would need rendering (transcripts with and without UTR, different strands, etc.).

File: src/scripts/TranscriptsTrack.js (hunk 4 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js]: Chevron blocks now only  are added for `blockStyle === 'UCSC-like'`.

Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js]: See commit 94a77d5

File: src/scripts/TranscriptsTrack.js (hunk 7 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js:2706]: I added a `colorBoxplotLabel` track option in commit 85cb079, which is a Boolean flag to render labels with a color background that matches the boxplot element coloring. 

The default setting for this option is `false`. If the flag is `true`, the `renderDirectionlessTextBg` function is used to render this background. The label text color will change based on the boxplot element coloring.

File: src/scripts/TranscriptsTrack.js (hunk 9 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js:3142]: I haven't run into a situation with more than one higlass display for my apps, which is likely why I did it this way. 

Is this something that Pete or Fritz might know more about?

**Comment by alexpreynolds** [comment]
I think I resolved the concerns outlined here, except for the question about how to get the "current" higlass container.

I may add some more commits from performance testing, as it does a bit more work than the generic gene annotations track and so slows down rendering a bit more than we were hoping.

Still, if there are any outstanding issues with what is there currently, please let me know!

File: src/scripts/TranscriptsTrack.js (hunk 3 above)
Comment (alexpreynolds) [reply src/scripts/TranscriptsTrack.js]: (Please see comments below)

**Comment by alexander-veit** [comment]
I am having some issues with the latest changes. Amino acids are not shown anymore when zoomed in and transcripts vanish occasionally when zooming. See here: https://gyazo.com/1a0e0d138845384ed4bce6b3c05088d0

**Comment by alexpreynolds** [comment]
I'm not entirely sure what to look for. Do you have a viewconf I can use to try to replicate this locally?

**Comment by alexander-veit** [comment]
You are not seeing the flickering issue locally? I used the transcript track configuration from https://higlass.io/app/?config=Yzld015XRFaCfCIYbB2MZQ

Edit: but you are seeing the issue in the gif above, right?

**Comment by alexpreynolds** [comment]
I'll try to set up a site to load your viewconf to try to reproduce things locally.

I did not see flickering in the GIF you posted. It otherwise looks normal to me, but the amino acid view is not my typical use case. 

Is there an easy way to repeat the same animation, but with the stock version of the plugin ("expected" vs "observed")? If not, that's fine.

**Pull request closed without merging**